### ISSUE TITLE
To test: various mathematical operations

### ISSUE THREAD

**Issue by ariasanovsky** [ariasanovsky/azdopt#43]
Test the following:

- [x] when assigning $p_\theta(s,\cdot)$ to a node, rescale by $\sum_{a\in\mathcal{A}(s)}p_\theta(s, a)$ before adding to the MCTS tree
- [x] when assigning $p_\theta(s, \cdot)$ to a root node with $\text{Dir}(\alpha)$, normalize before averaging
  - ~~[ ] make an Issue for the following question:~~
    - ~~Can we implement a different `StateActionData` that uses a fixed-size array and a priority queue?~~
- [x] when minimizing cross-entropy loss, mask probabilities or logits
  - this avoid punishing positive probability mass on invalid actions
  - is it also correct to apply this mask when inserting nodes into the tree?
    - [x] create a method to mask probabilities when inserting nodes
- ~~[ ] include $\overline{g}_\theta^\ast(s_t)$ in `StateActionData` to improve root node selection~~
- ~~[ ] can we replace each value of $n(s, a)$ with $n(sa)$ during calculation of $u(s, a)$?~~
  - Recall that $u(s, a)$ is calculated during `simulate_once` and would require simultaneously viewing:
    - `self.data.iter_mut()` and `self.data.iter().skip(1)`, which is sound, but not easy to do
  - Also, this may lead to more conservative estimates for $u(s, a)$ inconsistent a MCTS rollout rooted at $s$
    - Intuitively, these estimates seem like a good tradeoff, and $u(s, a)$ is an arbitrary heuristic anyway
- [x] not mathematical, but reuse the allocation from the `INTMinTree` instead of consuming when selecting the next node

**Comment by ariasanovsky** [comment]
For the $n(s, a)\to n(sa)$ trick, we somehow would need to replace

```rust
for data in data.iter_mut() {
```

with an iterator with items with type `(&mut T, Option<&T>)` (e.g., mutate the current element, but read from the next element if it exists.

**Comment by ariasanovsky** [comment]
Also, this could cascade a change to to the `INTActionData` struct. Perhaps more values get stored on nodes instead of transitions. I do like the idea of altering the $$p(s, a)\cdot\dfrac{f(n(s))}{1+n(s, a)}$$ term to use $n(sa)$ since this is the **exploration** term, but leaving the $q(s, a)$ term in-tact, otherwise it loses its tangible meaning as an average.

**Comment by ariasanovsky** [comment]
> Also, this could cascade a change to to the `INTActionData` struct. Perhaps more values get stored on nodes instead of transitions. I do like the idea of altering the p(s,a)⋅f(n(s))1+n(s,a) term to use $n(sa)$ since this is the **exploration** term, but leaving the $q(s,a)$ term in-tact, otherwise it loses its tangible meaning as an average.

Expanding on this idea, this would:
- require a small refactor from the `self.data.iter_mut()` so that it also peeks ahead at the remaining array (doable)
- and require a `BTreeMap` lookup for every single visited *and unvisited* action

I like this idea less. If it only required the lookup for visited actions, that's perfectly fine. But these spaces have many possible actions. Most likely I will nix this idea.

**Comment by ariasanovsky** [comment]
Significant improvement with logit mask applied to cross entropy loss calculation.
**Note**: The oscillation is a result of an unoptimized policy for the optimizer and for the root selection.

![tensorboard plot where cross entropy loss oscillates between 4.5 and 5.2](https://github.com/ariasanovsky/azdopt/assets/14944402/ca030e43-15f1-4b48-b493-ea67cebefde9)
![tensorboard plot where cross entropy loss oscillates between 1.4 and 5](https://github.com/ariasanovsky/azdopt/assets/14944402/81e9ff3f-1e60-426e-be0d-418b6d55f150)

**Comment by ariasanovsky** [comment]
Closed with #73 #74 (refactored with $\mu$ zero algorithm
